### PR TITLE
Issue 107

### LINKED ISSUE #107: Raise DragoneyeException when creating aws sessions on all scenarios

### PR DESCRIPTION
**Pull request by Tomerva**
fix #107

### CODE REVIEW

File: dragoneye/cloud_scanner/aws/aws_session_factory.py [hunk 1]
```diff
@@ -20,7 +20,12 @@ def get_session(profile_name: Optional[str] = None, region: Optional[str] = None
         else:
             start_msg = f'{start_msg} using AWS auth-chain'
         logger.info(f'{start_msg}...')
-        session = boto3.Session(**session_data)
+        try:
+            session = boto3.Session(**session_data)
+        except ClientError as ex:
+            raise DragoneyeException(f'An exception happened while trying to create session: {str(ex)}', ex.response['Error']['Message'])
```
Comment (alonashkenazi) [inline dragoneye/cloud_scanner/aws/aws_session_factory.py:26]: is `ex.response['Error']['Message']` safe? maybe user get?

File: dragoneye/cloud_scanner/aws/aws_session_factory.py [hunk 2]
```diff
@@ -29,20 +34,25 @@ def get_session(profile_name: Optional[str] = None, region: Optional[str] = None
...
+            session_data = {
+                "aws_access_key_id": credentials['AccessKeyId'],
+                "aws_secret_access_key": credentials['SecretAccessKey'],
+                "aws_session_token": credentials['SessionToken']
+            }
+            if region:
+                session_data['region_name'] = region
+            session = boto3.Session(**session_data)
+        except ClientError as ex:
+            raise DragoneyeException(f'An exception happened while trying to assume role: {str(ex)}', ex.response['Error']['Message'])
```
Comment (alonashkenazi) [inline dragoneye/cloud_scanner/aws/aws_session_factory.py:53]: same comment

File: dragoneye/cloud_scanner/azure/azure_authorizer.py [hunk 3]
```diff
@@ -23,13 +25,17 @@ def get_authorization_token(tenant_id: Optional[str] = None,
...
         `az account get-access-token <https://docs.microsoft.com/en-us/cli/azure/account?view=azure-cli-latest#az_account_get_access_token>`__
         """
+        validate_uuid(subscription_id, 'Invalid subscription id')
         if not (client_id and client_secret and tenant_id):
             token = AzureAuthorizer._get_token_from_az_cli()
         else:
             token = AzureAuthorizer._get_token_from_credentials(tenant_id, client_id, client_secret)
 
+        auth_token = f'Bearer {token}'
+        AzureAuthorizer._test_connectivity(subscription_id, auth_token)
```
Comment (alonashkenazi) [inline dragoneye/cloud_scanner/azure/azure_authorizer.py:35]: why did you add test connectivity as part of the auth flow, shouldn't this be a separate function that the user of Dragon can use?

File: dragoneye/cloud_scanner/azure/azure_authorizer.py [hunk 4]
```diff
@@ -71,3 +77,14 @@ def _get_token_from_az_cli() -> str:
             ind = output.rindex('}') + 1
             output = output[:ind]
             return json.loads(output)['accessToken']
+
+    @staticmethod
+    def _test_connectivity(subscription_id, token):
```
Comment (alonashkenazi) [inline dragoneye/cloud_scanner/azure/azure_authorizer.py:82]: need to be public function (like AWS)

File: dragoneye/scan.py [hunk 5]
```diff
@@ -121,7 +121,7 @@ def azure(cloud_account_name: str,
         should_clean_before_scan=clean,
         output_path=output_path)
 
-    auth_header = AzureAuthorizer.get_authorization_token(tenant_id, client_id, client_secret)
+    auth_header = AzureAuthorizer.get_authorization_token(subscription_id, tenant_id, client_id, client_secret)
```
Comment (alonashkenazi) [inline dragoneye/scan.py:124]: tenant needs to be before subscription

**Review by alonashkenazi — CHANGES_REQUESTED** [review]

File: dragoneye/cloud_scanner/aws/aws_session_factory.py (hunk 1 above)
Comment (Tomerva) [reply dragoneye/cloud_scanner/aws/aws_session_factory.py:26]: yeah its part of ClientError

File: dragoneye/cloud_scanner/azure/azure_authorizer.py (hunk 3 above)
Comment (Tomerva) [reply dragoneye/cloud_scanner/azure/azure_authorizer.py:35]: basically i added this for 2 reasons:
1. To be able to check if subscription id is correct. The token can be generated successfully without using subscription id. however, the sub id might be wrong and then all scan commands will fail.
2. To be consistent - this way all scenarios raise exception if auth fails for some reason

File: dragoneye/scan.py (hunk 5 above)
Comment (Tomerva) [reply dragoneye/scan.py:124]: why is that?
i prefer it this way, because subscription id is mandatory, and the rest are not

**Review by alonashkenazi — APPROVED** [review]
